### ISSUE TITLE
Rework itemRenderer prop

### ISSUE THREAD

**Issue by Ilaiwi** [namespace-ee/react-calendar-timeline#289]
Hi, 

I am working on a schedule to manage employee shifts. The library till now is working really well (Thank you for the effort). But the main issues I am having is that we don't have access to render all of the items differently. So my code to manage Items and how they should look in different scenarios was split into two parts: First part is a custom `Item` component which handles the rendering of the inner div of the item and the class name property in the array `items` passed to the component which I find awkward.

I understand that you had to do that in order to guarantee that the item renders with the correct height and left values and handlers

My suggestion is based on the renderProp pattern that `react-beautiful-dnd` employes, where the library trusts the developer to give those key attributes to the parent `div` via an argument called provided. This will give developers who need to do highly custom item to do so.

please refer to [react beautiful dnd droppable](https://github.com/atlassian/react-beautiful-dnd#droppable) for an example

I think user code will be something like this: 

```
<Timeline groups={groups}
    items={items}
    defaultTimeStart={moment().add(-12, 'hour')}
    defaultTimeEnd={moment().add(12, 'hour')}
    fullItemRenderer={(item, provided, providedDragLeft, providedDragRight) => (
        <div
            ref={provided.getInnerRef}
            // this includes the key properties like `left` and `height` and handlers (drag, click... handlers)
            {...provided.itemProps}
        >
            <div
                ref={providedDragLeft.getInnerRef}
                {...providedDragLeft.itemProps}
            >
            </div>
                {item.title}
            <div
                ref={providedDragRight.getInnerRef}
                {...providedDragRight.itemProps}
            >
            </div>
        </div>
    )}
/>
```
of course this prop will be optionals and will replace `itemRenderer` when it only excists

P.S. I have the resources to develop this feature if you think it is a good idea.

**Comment by Ilaiwi** [comment]
@mcMickJuice Thank you! I will start with this very soon under prop name `fullItemRenderer` so maybe we don't push this as a breaking change for now?

**Comment by mcMickJuice** [comment]
No, I'd prefer we just replace itemRenderer and include instructions for migrating from the previous version.

**Comment by mcMickJuice** [comment]
Yikes, I'll look at that bug.  Continue working on your branch (off of develop) and we can merge afterward

**Comment by mcMickJuice** [comment]
@Ilaiwi fix should be on develop. please merge when you have a chance.

Thanks for pointing this out!

**Comment by Ilaiwi** [comment]
@mcMickJuice Hi, I implemented the render prop with prop getter pattern to render the items and made the default prop the previous render function but I had problems wasn't as snappy when moving items as the current version of master (0.16v). I also tried the provide patterns since it has fewer function calls but it was slower as well. Please Take a look at [my fork](https://github.com/Ilaiwi/react-calendar-timeline/tree/itemRenderer). Please, let me know if you have any idea on how to maintain the performance or your thoughts on the performance hit.

**Comment by mcMickJuice** [comment]
Great! Can you open a PR against master so I can review?

**Comment by Ilaiwi** [comment]
@mcMickJuice to be clear with only the optimizations?

**Comment by mcMickJuice** [comment]
Can you open a PR for the itemRenderer prop reworking only - [this branch](https://github.com/Ilaiwi/react-calendar-timeline/tree/itemRenderer)? Not with the optimizations?  It's good practice to keep PRs very specific.

We can discuss the optimizations at a later point in time.

**Comment by mcMickJuice** [comment]
This has been published to @beta .  I'm going to merge other substantial changes first then publish to 0.19.0.  Also gotta add a migration from existing item renderer and this one.

Thanks @Ilaiwi !

**Comment by mcMickJuice** [comment]
Released in 0.19.0
